app/tests/test_keycloak_manager.py

```diff
@@ -0,0 +1,160 @@
+from unittest.mock import patch
+
+
+@patch(
+    "app.main.authorize.access_token_sign_in_required.keycloak.KeycloakOpenID.introspect"
+)
+def test_decode_token(mock_decode_keycloak_access_token, app):
+    """
+    Given an active access token in the session,
+    When access token given
+    Then it should decode token
+    """
+    mock_decode_keycloak_access_token.return_value = {
+        "active": True,
+        "groups": ["application_1/foo", "application_2/bar"],
+    }
+    app.config["FORCE_AUTHENTICATION_FOR_IN_TESTING"] = True
+    with app.test_client() as client:
+        with client.session_transaction() as session:
+            session["access_token"] = "valid_token"
+
+        assert mock_decode_keycloak_access_token.return_value["active"]
+        assert (
+            mock_decode_keycloak_access_token.return_value["groups"][0]
+            == "application_1/foo"
+        )
+
+
+@patch(
+    "app.main.authorize.access_token_sign_in_required.keycloak.KeycloakOpenID.introspect"
+)
+def test_get_user_transferring_body_groups_invalid_access_token(
+    mock_decode_keycloak_access_token, app
+):
+    """
+    Given an active access token in the session,
+    When access token given
+    Then it should not return user groups
+    """
+    mock_decode_keycloak_access_token.return_value = {}
+    user_groups_list = []
+    app.config["FORCE_AUTHENTICATION_FOR_IN_TESTING"] = True
+    with app.test_client() as client:
+        with client.session_transaction() as session:
+            session["access_token"] = None
+
+        assert len(user_groups_list) == 0
```

Comment:
this isnt testing anything useful, since you define `user_groups_list = []` on line 41 and then assert the length of this list to be 0.

Also the name of the test and the docstring dont seem to match up.

Going by the test name, I assume you wanted to assert what we expect when we pass an invalid access token (determined by mocking the introspect call to keycloak) to `get_user_transferring_body_groups`

And actually, I just realised that none of these tests are calling `get_user_transferring_body_groups` and are all just testing that `assert len([]) == 0`

apart from the last one which tests the equialvent of `assert len(["a", "b"]) == 2`

Let me know if you want a hand with these

Reply:
test cases has been modified and a new test with no token has been added

Reply:
all the tests still dont look right, will hop on a call, will be easier